lib/elixir/lib/enum.ex

```diff
@@ -711,8 +771,8 @@ defmodule Enum do
   end
 
   def empty?(enumerable) do
-    case Enumerable.count(enumerable) do
-      {:ok, value} when is_integer(value) ->
+    case Enumerable.slice(enumerable) do
+      {:ok, value, _} ->
```

Comment:
According to documentation I should implement `count/1` if I can count faster than O(n) but only return `{:ok, count, _}` from `slice/1` if I can count in constant time. So by changing to `slice` here, aren't we saying "if you can count really fast we'll take that but if you can count better than O(n) we won't use that"?

Reply:
I did this change because the only example I can think of where slice and count differs is file stream and in file streams it is cheaper to go the slice route than counting all new lines in a file just to compare it to zero.

Reply:
@whatyouhide other than that, is the difference between count and slice clear? What is your feeling about having both `slice` and `count` returning a count? Does it feel like it is duplicated functionality?

To provide another example besides File Stream of where count could be implemented but not slice, it would be if we implemented Enumerable for strings graphemes. There we can count without emitting the intermediate binaries by simply computing the graphemes size. It is still linear but we would save in allocations.

Reply:
I wouldn't say it's superclear I feel. I understand why there are both but it's a thin difference. However I am not sure how we could avoid having the two of them without sacrificing something.

Reply:
I don't know if it's driving away of what you want for the implementation, but I see that the code in `slice` and `count` are similar for counting. You could take the idea that you developed in `slice` and use it in `count`, asking an implementation that returns a value if it's O(1), but that returns a function if it's expensive. Them you could use them separated and decide what is better, I don't know if it makes sense because the uses are complex.

Reply:
Of course, `count` needs another name, the actual count would just call this function

Reply:
@ggcampinho that would definitely work if we want to merge `slice` and `count`. :+1: On the other hand it could make the return type of `slice` more complex and increase the chances someone would return the wrong value for the `size` in slice (i.e. they return an integer but it should have been a fun).

Reply:
@ggcampinho your suggestion makes me realize that merging them is probably not worth it. So the only remaining question is if we want to allow count to still be implemented.

And the other concern is that we saw some incorrect implementations of `count` when exploring GitHub. Those implementations used linear or expensive implementations of `count` which could actually make many Enum operations slower. We even did this mistake in Elixir itself. If we remove `count` and keep only `slice`, will those mistakes be easier or harder to repeat compared to if we had both?

Reply:
@josevalim  I meant extracting from `slice` and moving to `count`, but a more complex `count`. I do agree with you that is better to have a lean contract and a new extra complex function can lead to problems, so I'm okay with the final decision.